src/Text.mo

```diff
@@ -26,30 +45,521 @@ module {
+                     i += 1;
+                   };
+                   case null {
+                     return #empty (take(i, p.chars()));
+                   }
+                 }
+               };
+               case null { return #success };
+             }
+           }
+         }
+       }
+     }
+  };
+
+  private class CharBuffer(cs : Iter.Iter<Char>) : Iter.Iter<Char> = {
+
+    var buff : Iter.Iter<Char> = empty();
+    var char : ?Char = null;
+
+    public func pushBack(cs0: Iter.Iter<Char>, c : Char) {
+       buff := cs0;
+       char := ?c;
+    };
+
+    public func next() : ?Char {
+      switch (buff.next()) {
+        case null {
+          switch char {
+            case (?c) {
+              char := null;
+              return ?c;
+            };
+            case null {
+              return cs.next();
+            };
+          }
+        };
+        case oc oc;
+      }
+    };
+  };
+
+  /// Returns the sequence of fields in `t`, derived from start to end,
+  /// separated by text matching [pattern](#type.Pattern) `p`.
+  /// Two fields are separated by exactly one match.
+  public func split(t : Text, p : Pattern) : Iter.Iter<Text> {
+    let match = matchOfPattern p;
+    let cs = CharBuffer(t.chars());
+    var state = 0;
+    var field = "";
+    object {
+      public func next() : ?Text {
+        switch state {
+          case (0 or 1) {
+            loop {
+              switch (match(cs)) {
+                case (#success) {
+                  let r = field;
+                  field := "";
+                  state := 1;
+                  return ?r
+                };
+                case (#empty cs1) {
+                  for (c in cs1) {
+                    field #= fromChar c;
+                  };
+                  let r =
+                    if (state == 0 and field == "")
+                      null
+                    else ?field;
+                  state := 2;
+                  return r;
+                };
+                case (#fail (cs1, c)) {
+                  cs.pushBack(cs1,c);
+                  switch (cs.next()) {
+                    case (?ci) {
+                      field #= fromChar ci;
+                    };
+                    case null {
+                      let r =
+                         if (state == 0 and field == "")
+                           null
+                         else ?field;
+                      state := 2;
+                      return r;
+                    }
+                  }
+                }
+              }
+            }
+          };
+          case _ { return null };
+        }
+      }
+    }
+  };
+
+  /// Returns the sequence of tokens in `t`, derived from start to end.
+  /// A _token_ is a non-empty maximal subsequence of `t` not containing a match for [pattern](#type.Pattern) `p`.
+  /// Two tokens may be separated by one or more matches of `p`.
+  public func tokens(t : Text, p : Pattern) : Iter.Iter<Text> {
```

Comment:
Should we be careful about using `token`?

Reply:
> Should we be careful about using token?

Why? (it's not reserved, if that's what you were thinking.)

Reply:
I assume you're joking but, if anything, should probably be called 'tokenize' since related  SML Basis 'fields' function got renamed from noun to verb split.

Reply:
OK with tokenize? Or maybe 'lex' ?

Reply:
I think `tokens` is a good name.

Reply:
@matthewhammer I think Stanley was referencing our blockchain denial

Reply:
That's worse, split is pretty standard.